TESTS/mbed_hal/i2c/main.cpp

```diff
@@ -0,0 +1,351 @@
+void test_i2c_start()
+{
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+
+    i2c_start(&obj);
+
+    i2c_free(&obj);
+}
+
+// Test that i2c_stop successfully sends stop signal to the bus
+void test_i2c_stop()
+{
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+
+    i2c_stop(&obj);
+
+    i2c_free(&obj);
+}
+
+// Test i2c_read/i2c_write calls for non-existent slave address
+void test_i2c_read_write()
+{
+    int ret;
+    const uint32_t size = 3;
+    uint8_t data[size];
+
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+
+    ret = i2c_write(&obj, MAKE_7BIT_WRITE_ADDRESS(I2C_7BIT_ADDRESS), (void *)data, size, false);
+    TEST_ASSERT_EQUAL(I2C_ERROR_NO_SLAVE, ret);
+
+    ret = i2c_write(&obj, MAKE_7BIT_WRITE_ADDRESS(I2C_7BIT_ADDRESS), (void *)data, size, true);
+    TEST_ASSERT_EQUAL(I2C_ERROR_NO_SLAVE, ret);
+
+    ret = i2c_read(&obj, MAKE_7BIT_READ_ADDRESS(I2C_7BIT_ADDRESS), (void *)data, size, false);
+    TEST_ASSERT_EQUAL(I2C_ERROR_NO_SLAVE, ret);
+
+    ret = i2c_read(&obj, MAKE_7BIT_READ_ADDRESS(I2C_7BIT_ADDRESS), (void *)data, size, true);
+    TEST_ASSERT_EQUAL(I2C_ERROR_NO_SLAVE, ret);
+
+    i2c_free(&obj);
+}
+
+#if DEVICE_I2CSLAVE
+// Test that i2c_slave_address successfully sets given address
+void test_i2c_slave_address()
+{
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_SLAVE);
+
+    i2c_slave_address(&obj, MAKE_7BIT_SLAVE_ADDRESS(I2C_7BIT_ADDRESS_MIN));
+    i2c_slave_address(&obj, MAKE_7BIT_SLAVE_ADDRESS(I2C_7BIT_ADDRESS));
+    i2c_slave_address(&obj, MAKE_7BIT_SLAVE_ADDRESS(I2C_7BIT_ADDRESS_MAX));
+
+    i2c_free(&obj);
+}
+
+// Test that i2c_slave_address successfully sets given address
+void test_i2c_slave_status()
+{
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_SLAVE);
+
+    i2c_slave_address(&obj, MAKE_7BIT_SLAVE_ADDRESS(I2C_7BIT_ADDRESS));
+
+    i2c_slave_status_t status = i2c_slave_status(&obj);
+    TEST_ASSERT_EQUAL(NoData, status);
+
+    i2c_free(&obj);
+}
+#endif
+
+#if DEVICE_I2C_ASYNCH
+volatile bool async_handler_called;
+void i2c_async_handler(i2c_t *obj, i2c_async_event_t *event, void *ctx)
+{
+    i2c_async_event_t *e = (i2c_async_event_t *)ctx;
+    e->sent_bytes = event->sent_bytes;
+    e->received_bytes = event->received_bytes;
+    e->error = event->error;
+    async_handler_called = true;
+}
+
+// Test i2c_transfer_async call for non-existent slave address
+void test_i2c_transfer_async()
+{
+    int ret;
+    const uint32_t size = 3;
+    uint8_t tx_data[size];
+    uint8_t rx_data[size];
+
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+
+    {
+        i2c_async_event_t event = {};
+        async_handler_called = false;
+        i2c_transfer_async(&obj, tx_data, size, NULL, 0, MAKE_7BIT_SLAVE_ADDRESS(I2C_7BIT_ADDRESS), true, i2c_async_handler, &event);
+        wait(1);
+        TEST_ASSERT_EQUAL(true, async_handler_called);
```

Comment:
`NUCLEO_F070RB` fails in this line. Shouldn't the `i2c_async_handler` be called here indicating an error?

Reply:
@scartmell-arm still failing here